tests/core/featurizers/test_single_state_featurizers.py

```diff
@@ -341,164 +301,269 @@ def test_single_state_featurizer_with_interpreter_state_with_action_listen(
+    tags_to_ids = f.entity_tag_specs[0].tags_to_ids
+    for idx, entity_tag in enumerate(entity_tags):
+        tags_to_ids[entity_tag] = idx + 1  # hence, city -> 1, city#to -> 2
+    assert sorted(list(encoded.keys())) == [ENTITY_TAGS]
+    assert np.all(
+        encoded[ENTITY_TAGS][0].features == [[0], [0], [0], [0], [1], [0], [2]]
+    )
 
 
-def test_state_features_for_attribute_raises_on_not_supported_attribute():
-    f = SingleStateFeaturizer()
+def test_encode_entities__with_bilou_entity_roles_and_groups():
 
-    with pytest.raises(ValueError):
-        f._state_features_for_attribute({}, "not-supported-attribute")
+    # Instantiate domain and configure the single state featurizer for this domain.
+    # Note that there are 2 entity tags here.
+    entity_tags = ["city", f"city{ENTITY_LABEL_SEPARATOR}to"]
+    domain = Domain(
+        intents=[],
+        entities=entity_tags,
+        slots=[],
+        responses={},
+        forms={},
+        action_names=[],
+    )
+    f = SingleStateFeaturizer()
+    f.prepare_for_training(domain, bilou_tagging=True)
 
+    # (1) example with both entities
 
-def test_to_sparse_sentence_features():
-    features = [
-        Features(
-            scipy.sparse.csr_matrix(np.random.randint(5, size=(5, 10))),
-            FEATURE_TYPE_SEQUENCE,
-            TEXT,
-            "some-featurizer",
-        )
+    # create message that has been tokenized and where entities have been extracted
+    text = "I am flying from London to Paris"
+    tokens = [
+        Token(text=match.group(), start=match.start())
+        for match in re.finditer(r"\S+", text)
     ]
+    entities = [
+        {
+            ENTITY_ATTRIBUTE_TYPE: entity_tags[0],
+            ENTITY_ATTRIBUTE_VALUE: "London",
+            ENTITY_ATTRIBUTE_START: 17,
+            ENTITY_ATTRIBUTE_END: 23,
+        },
+        {
+            ENTITY_ATTRIBUTE_TYPE: entity_tags[1],
+            ENTITY_ATTRIBUTE_VALUE: "Paris",
+            ENTITY_ATTRIBUTE_START: 27,
+            ENTITY_ATTRIBUTE_END: 32,
+        },
+    ]
+    message = Message({TEXT: text, TOKENS_NAMES[TEXT]: tokens, ENTITIES: entities})
 
-    sentence_features = SingleStateFeaturizer._to_sparse_sentence_features(features)
+    # create a lookup table that has seen this message
+    lookup_table = E2ELookupTable()
+    lookup_table.add(message)
 
-    assert len(sentence_features) == 1
-    assert FEATURE_TYPE_SENTENCE == sentence_features[0].type
-    assert features[0].origin == sentence_features[0].origin
-    assert features[0].attribute == sentence_features[0].attribute
-    assert sentence_features[0].features.shape == (1, 10)
+    # encode!
+    encoded = f.encode_entities(
+        {TEXT: text, ENTITIES: entities,},
+        e2e_features=lookup_table,
+        bilou_tagging=True,
+    )
+    assert sorted(list(encoded.keys())) == sorted([ENTITY_TAGS])
+    assert np.all(
+        encoded[ENTITY_TAGS][0].features == [[0], [0], [0], [0], [4], [0], [8]]
```

Comment:
Could you please add a comment explaining where the `[[0], [0], [0], [0], [4], [0], [8]]` comes from?

Reply:
I'd need to look into the bilou tagging spagetti to find out and I've already cleaned this up much more than was needed - so I won't add this now. We can pick this up again when doing the disentangling (and/or when creating a proper entities abstraction 😬 )